document/core/exec/instructions.rst

```diff
@@ -2680,14 +2680,40 @@ Control Instructions
 :math:`\RETHROW~l`
 ..................
 
+1. Assert: due to :ref:`validation <valid-rethrow>`, the stack contains at least :math:`l+1` labels.
+
+2. Let :math:`L` be the :math:`l`-th label appearing on the stack, starting from the top and counting from zero.
+
+3. Assert: due to :ref:`validation <valid-rethrow>`, :math:`L` is a catch label, i.e., a label of the form :math:`(\LCATCH~[t^\ast])`.
+
+4. Repeat :math:`l+1` times:
+
+   a. While the top of the stack is a value, a |handler|, or a |CAUGHTadm| instruction, do:
+
+      i. Pop the instruction from the stack.
+
+   b. Assert: due to :ref:`validation <valid-br>`, the top of the stack now is a label.
+
+   c. Pop the label from the stack.
+
+5. Assert: due to **TODO**, the last two instructions popped are a |CAUGHTadm| instruction and the :math:`l+1`-th label, in that order.
+
 .. todo::
-   Add prose for the |RETHROW| execution step.
+   Justify the existence of a |CAUGHTadm| here. Perhaps extend some typing rule?
+
+6. Let :math:`\CAUGHTadm\{a~\val^\ast\}` be the instruction after the label :math:`L`.
+
+7. Put all the popped values back on the stack, except the last instruction, the |RETHROW|.
```

Comment:
That seems really odd. As commented elsewhere, I think we should avoid this.

Reply:
How do you think can we avoid this? Can we use a concept of shadow stack or something, so we don't really pop the values in 4-a-i above but just scan through it?

Should we do this looping thing in 4-a-i after all? All other branch-like instructions do it so there must be a reason, but can't we just say, because of validation, we can assert `L`th is a catch label with type `[t*]->[]` and `L-1`th is a `caught` instruction because of try-catch execution rule, and just refer to the `val`s?

Reply:
As mentioned above, I just pushed an attempt to avoid all this pushing a popping. WDYT?